packages/runtime/src/graph.ts

```diff
@@ -20,12 +20,109 @@ export class Graph {
 		return new Graph({ id });
 	}
 
-	static async new(..._args: tg.Args<Graph.Arg>): Promise<Graph> {
-		return todo();
+	static async new(...args: tg.Args<Graph.Arg>): Promise<Graph> {
+		let arg = await Graph.arg(...args);
+		let nodes = await Promise.all(
+			(arg.nodes ?? []).map(async (node) => {
+				if (node.kind === "directory") {
+					return {
+						kind: "directory" as const,
+						entries: node.entries ?? {},
+					};
+				} else if (node.kind === "file") {
+					return {
+						kind: "file" as const,
+						contents: await tg.blob(node.contents),
+						dependencies: node.dependencies ?? undefined,
+						executable: node.executable ?? false,
+					};
+				} else if (node.kind === "symlink") {
+					return {
+						kind: "symlink" as const,
+						artifact: node.artifact ?? undefined,
+						path: node.path !== undefined ? tg.path(node.path) : undefined,
+					};
+				} else {
+					return tg.unreachable(node);
+				}
+			}),
+		);
+		return new Graph({ object: { nodes } });
 	}
 
-	static async arg(..._args: tg.Args<Graph.Arg>): Promise<Graph.ArgObject> {
-		return todo();
+	static async arg(...args: tg.Args<Graph.Arg>): Promise<Graph.ArgObject> {
+		let resolved = await Promise.all(args.map(tg.resolve));
+		let flattened = flatten(resolved);
+		let nodes = [];
+		let offset = 0;
+		for (let arg of flattened) {
+			let argNodes = arg instanceof Graph ? await arg.nodes() : arg.nodes || [];
+			for (let argNode of argNodes) {
+				if (argNode.kind === "directory") {
+					let node: tg.Graph.DirectoryNodeArg = { kind: "directory" as const };
+					if ("entries" in argNode) {
+						if (argNode.entries !== undefined) {
+							node.entries = {};
+							for (let name in argNode.entries) {
+								if (typeof argNode.entries[name] === "number") {
+									node.entries[name] = argNode.entries[name] + offset;
+								}
```

Comment:
I think we also need the `else` branch for non-numbers, `node.entries[name] = argNode.entries[name]`